### PR TITLE
feat: basic structure based on cookiecutter

### PR DESCRIPTION
**Pull request by psankhe28**
#4 

## Summary by Sourcery

This pull request sets up the basic project structure using Cookiecutter, including Makefile, Sphinx documentation, GitHub Actions for CI/CD, a Code of Conduct, and initial test placeholders.

- **New Features**:
    - Introduced a Makefile with various targets for environment management, code quality, testing, and documentation generation.
    - Added Sphinx configuration for generating project documentation.
    - Implemented a GitHub Action for setting up Python and Poetry environments.
    - Created multiple GitHub workflows for code quality checks, testing, PR validation, documentation checks, and vulnerability scans.
    - Added a Code of Conduct document.
    - Included a PULL_REQUEST_TEMPLATE.md for standardizing pull request descriptions.
    - Added a .pre-commit-config.yaml for pre-commit hooks.
    - Introduced a main entry point for the tus_storagehandler module.
- **Enhancements**:
    - Configured Read the Docs with a .readthedocs.yaml file for building documentation.
    - Added a .yamllint.yaml file for YAML linting configuration.
- **CI**:
    - Set up GitHub Actions for code quality, testing, PR validation, documentation checks, and vulnerability scans.
- **Documentation**:
    - Added initial Sphinx documentation structure and configuration.
    - Included a Makefile and make.bat for building documentation.
- **Tests**:
    - Added placeholder test files for unit and integration tests to avoid errors during test runs.
- **Chores**:
    - Removed the README.md file.

### CODE REVIEW

File: pyproject.toml
Comment (uniqueg) [inline pyproject.toml:19]: Capitalize ELIXIR

File: LICENSE
Comment (uniqueg) [inline LICENSE:189]: Capitalize ELIXIR

File: README.md
Comment (uniqueg) [inline README.md:1]: I think there's no need to remove the existing `README.md`, so please re-add.

File: tus_storagehandler/main.py
Comment (uniqueg) [reply tus_storagehandler/main.py:6]: I think this is a good comment, please add logging.

**Review by uniqueg — CHANGES_REQUESTED** [review]
A few minor changes. Let's get this merged soon so that the actual coding can begin :rocket:

**Comment by uniqueg** [comment]
Oh, and please make sure all of the tests pass :pray:

File: pyproject.toml
Comment (JaeAeich) [inline pyproject.toml:61]: change it to 4 and run,

```bash
make v
source .venv/bin/activate
make fl
make pc
git add .
git commit -m "style: change indent"
```

PS: The above command are ment to tell you what to do if, they fail (like you use windows) then change the commds accordingly, you'll find the commands in Makefile.

File: pyproject.toml
Comment (JaeAeich) [inline pyproject.toml:26]: change it to `python = ">= 3.11<4.0.0"`, right @uniqueg?\

Because python is largely backwards compatible but if we leave it to python = "^3.11", and someone with global installation of python 3.12 tries to run the proj, poetry won't allow it.

File: README.md
Comment (JaeAeich) [reply README.md:1]: Also please refer to the latest commit of cookiecitter.

Maybe recreate the proj with the same config and just copy paste Readme.md, PULL_REQUEST_TEMPLATE, CODE_OF_CONDUCT and general-purpose.md. Well that is the fastest way :) than fixing.

File: .pre-commit-config.yaml
Comment (JaeAeich) [inline .pre-commit-config.yaml:28]: remove this line

File: .github/workflows/docs.yaml
Comment (JaeAeich) [inline .github/workflows/docs.yaml:1]: @uniqueg please review this workflow, before merge or either we should remove it from the cookiecutter too, this hasn't been tested.

File: tus_storagehandler/main.py
Comment (JaeAeich) [reply tus_storagehandler/main.py:6]: Yes agreed, but I think @psankhe28 should have selected "n" for `add_script` option while templating this proj, then this file wouldn't have been generated and nor will the console script would have been added, I don;t think we need a console script for this proj do we?

But if we do, then maybe do this in another PR(leave this to be templating only :man_shrugging: ), or I should add this in the cookiecutter itself.

PS: This is just an opinion tho :).

File: .cruft.json
Comment (JaeAeich) [reply .cruft.json:9]: PS: This has been changed to a list (sorry :face_in_clouds: ), so this should now be something like
{
...
  "commit": "43f9bd92eaf21557e87d3932524fe5ffb6fd7ef4",
  "checkout": null,
  "context": {
    "cookiecutter": {
      "author_name": "ELIXIR Cloud AAI",
      "author_email": "cloud-service@elixir-europe.org",
      "development_status": "1 - Planning", <- or 3 - Alpha
      "short_description": "A short description of the project.",
      "project_name": "python-cookiecutter",
      "project_slug": "python_cookiecutter",
....
}

File: .github/workflows/code_test.yaml
Comment (JaeAeich) [inline .github/workflows/code_test.yaml:14]: Change this to `Integration Test` please.

**Comment by JaeAeich** [comment]
@uniqueg test ci is failing because this repo doesn't have the token setup.
Please check the comment on `api ci`.

File: pyproject.toml
Comment (uniqueg) [reply pyproject.toml:26]: In fact, there likely won't be a Python 4 - or at least not within the next couple of years. At least that's what I'm reading.

File: .github/workflows/docs.yaml
Comment (uniqueg) [reply .github/workflows/docs.yaml:1]: Looks good to me. Let's run it and see if it gives any errors. If so, we can update it until it works, then also change it in the cookiecutter.

**Comment by uniqueg** [comment]
> @uniqueg test ci is failing because this repo doesn't have the token setup.
> Please check the comment on `api ci`.

Okay, I'll check that later.

**Comment by uniqueg** [comment]
> @uniqueg test ci is failing because this repo doesn't have the token setup. Please check the comment on `api ci`.

I have added the PyPI and Codecov tokens. For the GH token, please specify exactly what permissions are needed, @JaeAeich (in the Cookiecutter repo) - I will add it afterwards.

File: tus_storagehandler/main.py
Comment (psankhe28) [reply tus_storagehandler/main.py:6]: @JaeAeich I had used the default value for `add_script`.

File: .cruft.json
Comment (uniqueg) [reply .cruft.json:9]: No, I think you should use "1 - Planning", including the number. As you progress, you can change this to "2 - Pre-Alpha" and "3 - Alpha" as we see fit.

File: pyproject.toml
Comment (psankhe28) [reply pyproject.toml:61]: In readme, according to the make fl command, following changes should be done:
`
error: `mke` should be `make`
  --> .\README.md:95:1
   |
95 | mke = 'mke'
   | ^^^
   |
error: `mke` should be `make`
  --> .\README.md:95:8
   |
95 | mke = 'mke'
   |        ^^^
   `
   
   Should I do this?

File: .cruft.json
Comment (psankhe28) [reply .cruft.json]: Understood, thanks!

**Comment by JaeAeich** [comment]
> I have added the PyPI and Codecov tokens.

Hey checkout test CI logs, it still shows that codecov token are missing.

File: .github/ISSUE_TEMPLATE/general-purpose.md
Comment (JaeAeich) [inline .github/ISSUE_TEMPLATE/general-purpose.md:32]: Remove these line, because there is nothing written in them, md formatter does this. IDK maybe its a bug of mdformatter.

**Comment by JaeAeich** [comment]
@uniqueg seems like I don't have permission to this repo, can you please add required permissions.

>❯ git push origin fix
remote: Permission to elixir-cloud-aai/tus-storagehandler.git denied to JaeAeich.
fatal: unable to access 'https://github.com/elixir-cloud-aai/tus-storagehandler.git/': The requested URL returned error: 403

**Comment by uniqueg** [comment]
> @uniqueg seems like I don't have permission to this repo, can you please add required permissions.
> 
> > ❯ git push origin fix
> > remote: Permission to elixir-cloud-aai/tus-storagehandler.git denied to JaeAeich.
> > fatal: unable to access 'https://github.com/elixir-cloud-aai/tus-storagehandler.git/': The requested URL returned error: 403

Done, @JaeAeich

File: pyproject.toml
Comment (uniqueg) [reply pyproject.toml]: @JaeAeich?

**Comment by uniqueg** [comment]
@JaeAeich: I think it makes much more sense if you review this PR as long as the Cookiecutter PR is still not merged, as it might require another round of back and forth with that one (which I'm already reviewing). Also, you are more familiar with the code and potential pitfalls.

Is that okay? Let's ideally get this done this week (both the Cookiecutter PR and then also this one), so that @psankhe28 gets unblocked.

**Comment by uniqueg** [comment]
Upon reviewing #1 and #9, I suppose they will also become redundant with the Cookiecutter template in place and with this PR updated, right, @JaeAeich?

It would be good to get to a point where we have only one of these PRs (appropriately named, perhaps `chose: initialize project` or similar), tests pass and it's okay on your end, @psankhe28 and @JaeAeich. I can then give it a final view before merging.

**Comment by uniqueg** [comment]
Closed in favor of #10

**Pull request closed without merging**
